base/ca/src/org/dogtagpki/ct/CTEngine.java

```diff
@@ -0,0 +1,745 @@
+
+                    int ct_port = ls.getUrl().getPort();
+                    logger.debug(method + "Log server port: " + ct_port);
+
+                    // TODO: Refactor to form right rest API
+                    String ct_uri = ls.getUrl() + "ct/v1/add-pre-chain";
+                    logger.debug(method + "Log server URI: " + ct_uri);
+
+                    String respS = certTransSendReq(ct_host, ct_port, ct_uri, ctRequest);
+                    if (respS == null) {
+                        errMsg = method + "Response from CT log server null";
+                        logger.warn(errMsg);
+                        // allow for CT log to fail to respond
+                        // skip to next CT log server
+                        continue;
+                    }
+                    logger.debug(method + "Response from CT log server " + respS);
+
+                    // verify the sct
+
+                    /* TODO this should be a configurable; hardcoded for now */
+                    boolean allowFailedSCTVerification = true;
+
+                    boolean verified = verifySCT(CTResponse.fromJSON(respS), tbsCert, ls.getPublicKey(), ctCA);
+                    if (verified) {
+                        logger.info(method + "verifySCT returned true; SCT is valid");
+                    } else {
+                        // log at WARN if !verified, regardless of how we are treating
+                        // failed verifications, because it is indicative of log server
+                        // misbehavoiur
+                        logger.warn(method + "verifySCT returns false; SCT failed to verify");
+                    }
+                    if (verified || allowFailedSCTVerification) {
+                        ctResponses.add(respS);
+                    } else {
+                        throw new EBaseException(errMsg);
+                    }
+                }
+
+                /**
+                 * Now onto turning the precert into a real cert with the SCT list extension
+                 */
+
+                // create SCT extension
+                Extension sctExt = createSCTextension(ctResponses);
+                if (sctExt == null) {
+                    errMsg = " createSCTextension returns null";
+                    logger.debug(method + errMsg);
+                    throw new EBaseException(errMsg);
+                }
+
+                // add the SCT extension
+                exts.set(sctExt.getExtensionId().toString(), sctExt);
+
+                certi.delete(X509CertInfo.EXTENSIONS);
+                certi.set(X509CertInfo.EXTENSIONS, exts);
+                printExtensions(exts);
+            }
+        } catch (Exception e) {
+            logger.error(method + "Error occurred: " + e.getMessage(), e);
+            throw new EBaseException(e.getMessage());
+        }
+    }
+
+    /*
+     * for debugging
+     */
+    protected void printExtensions(CertificateExtensions exts) {
```

Comment:
Minor: This should be moved into a separate Class of its own (probably `DebuggingUtils` and make it a static method. This is not CT specific, so probably to `com.netscape.certsrv.util`, as this can be reused by other tools as well

Reply:
sounds good.